README.md

```diff
@@ -1,37 +1,264 @@
+additionally generate a controller per type.  The interface to the controller can be seen in the [Looking at Structures Used in Wrangler](#a-look-at-structures-used-in-wrangler) section.
+
+<br>
+
+The controller type along with other helpful types and functions are provided by another project [lasso](https://github.com/rancher/lasso). Lasso ties together the aforementioned types while wrangler leverages them in a user friendly way.
+
+To use the controller to run custom code for Kubernetes types all one needs to do is register OnChange handlers and run the controller.  Also using the controller interface one can access the client and caches through a simple flat API.
+
+A typical, non-wrangler Kubernetes application would most likely use an informer for a type to add an event handler. Instead, wrangler uses lasso to register each handler which then aggregates the handlers into one function that accepts an object for the controller's type and then runs that object through all the handlers. This function is then registered to the Kubernetes informer for that controller's respecctive type. This is done, presumedly, so that an object can run through the handlers in a serialized way. This allows each handler to receive the updated version of the object and avoid many conflicts that would otherwise occur if the handlers were not chained together in this fashion.
+
+<br>
+
+## Useful Definitions:
+### factory:
+Controllers are categorized by their API group and bundled into a struct called a factory.
+
+### informers:
+Broadcasts events for a given type and can register handlers for those events.
+
+### listers:
+Sometimes referred to as a cache, uses informers to update a local list of a type of object to avoid making requests to the K8s API
+
+### event handlers:
+Functions that run when a particular event is applied to the type of resource the event handler is assigned to.
+
+### workqueue:
+A queue of items to be proccessed. In the context being discussed this will usually be a queue of objects of a certain resource type waiting to be processed by all handlers assigned to that resource type.
+
+<br>
+
+# How to Use Wrangler
+
+Generate controllers for CRD's by using Run() from the controllergen package. This will look like the
+following:
+
+```golang
+controllergen.Run(args.Options{
+		OutputPackage: "github.com/rancher/rancher/pkg/generated",
+		Boilerplate:   "scripts/boilerplate.go.txt",
+		Groups: map[string]args.Group{
+			"management.cattle.io": {
+				PackageName: "management.cattle.io",
+				Types: []interface{}{
+					// All structs with an embedded ObjectMeta field will be picked up
+					"./pkg/apis/management.cattle.io/v3",
+					v3.ProjectCatalog{},
+					v3.ClusterCatalog{},
+				},
+				GenerateTypes: true,
+			},
+			"ui.cattle.io": {
+				PackageName: "ui.cattle.io",
+				Types: []interface{}{
+					"./pkg/apis/ui.cattle.io/v1",
+				},
+				GenerateTypes: true,
+			},
+		},
+	})
+```
```

Comment:
Do you mean in options or the catalog structs?

Reply:
Maybe both? comments for the args.Options, but now I am curious about where the embedded Catalog structs are defined. Were they generated beforehand somehow? Or can these structs be any struct that has an embedded k8s.io/apimachinery/pkg/apis/meta/v1.ObjectMeta struct?
```suggestion
```golang
controllergen.Run(args.Options{
                 // Location to store generated packages.
		OutputPackage: "github.com/rancher/rancher/pkg/generated",
		// Whatever Boilerplate is.
		Boilerplate:   "scripts/boilerplate.go.txt",
		// Things you want to generate. Is the key always the APIGroup.
		Groups: map[string]args.Group{
			"management.cattle.io": {
			         // Does this have to match the APIGroup?
				PackageName: "management.cattle.io",
				// List of string package names or structs? can they be pointers? 
				Types: []interface{}{
					// All structs with an embedded ObjectMeta field will be picked up
					"./pkg/apis/management.cattle.io/v3",
					v3.ProjectCatalog{},
					v3.ClusterCatalog{},
				},
				// What gets generated if this is false?
				GenerateTypes: true,
			},
			"ui.cattle.io": {
				PackageName: "ui.cattle.io",
				Types: []interface{}{
					"./pkg/apis/ui.cattle.io/v1",
				},
				GenerateTypes: true,
			},
		},
	})
```
```

Reply:
It's the latter. The Catalog is generated above since it is in the named package and directly has the ObjectMeta field. The ClusterCatalog and ProjectCatalog types are explitly named because they don't directly have the ObjectMeta field and instead embed a struct that does. They are both basically namespaced versions of the Catalog type. I have added explanations for this and provided an additional example for a simpler type.